src/laser_scan_assembler.cpp

```diff
@@ -52,32 +62,212 @@ class LaserScanAssembler : public BaseAssembler<sensor_msgs::LaserScan>
 public:
   LaserScanAssembler() : BaseAssembler<sensor_msgs::LaserScan>("max_scans"), filter_chain_("sensor_msgs::LaserScan")
   {
+    ROS_INFO("Constructor of LaserScanAssembler");
     // ***** Set Laser Projection Method *****
     private_ns_.param("ignore_laser_skew", ignore_laser_skew_, true);
 
+    // Automatically subscribe to scans or wait for start-service?
+    private_ns_.param("subscribe_directly", subscribe_directly_, true);
+
     // configure the filter chain from the parameter server
     filter_chain_.configure("filters", private_ns_);
 
+    start_collection_server_ = n_.advertiseService("start_collection", &LaserScanAssembler::startCollection, this);
+    stop_collection_server_ = n_.advertiseService("stop_collection_and_assemble_scans2", &LaserScanAssembler::stopCollectionAndAssembleScans2, this);
+    
+    pointcloud2_pub_ = n_.advertise<sensor_msgs::PointCloud2>("depth", 1);
+    stretched_range_image_pub_ =  n_.advertise<sensor_msgs::Image> ("range_image", 1);
+    stretched_depth_image_pub_ =  n_.advertise<sensor_msgs::Image> ("depth_image", 1);
+
+    if (subscribe_directly_)
+      subscribe();
+
+  }
+
+  ~LaserScanAssembler()
+  {
+
+  }
+
+  void subscribe()
+  {
     // Have different callbacks, depending on whether or not we want to ignore laser skews.
     if (ignore_laser_skew_)
+    {
       start("scan");
+    }
     else
     {
       start();
       skew_scan_sub_ = n_.subscribe("scan", 10, &LaserScanAssembler::scanCallback, this);
     }
+    ROS_INFO("Started listening to scans");
   }
 
-  ~LaserScanAssembler()
+  void unsubscribe()
+  {
+    ROS_INFO("Stopping to listen to scans");
+    if (ignore_laser_skew_)
+    {
+      scan_sub_.unsubscribe();
+    }
+    else
+    {
+      skew_scan_sub_.shutdown();
+    }
+    ROS_INFO("Stopped listening to scans");
+  }
+
+  cv::Mat sortMatRowBy(cv::Mat key, cv::Mat unsorted, std::vector<double>& sorted_keys, std::multimap<double, std::size_t>& reordering)
+  {
+    ROS_DEBUG_STREAM("key.size(): " << key.size() << ", unsorted.size(): " << unsorted.size());
+    cv::Mat sorted = cv::Mat(unsorted.size(), unsorted.type());
+
+    // A multimap inserts elements in order (per https://stackoverflow.com/questions/1577475/c-sorting-and-keeping-track-of-indexes)
+    std::multimap<double, std::size_t> remapping;
+    for (uint origIndex = 0; origIndex != key.size[0]; ++origIndex)
+    {
+        reordering.insert({key.at<double>(origIndex), origIndex});
+    }
+
+    sorted_keys = std::vector<double>();
+    for (auto it=reordering.begin(); it!=reordering.end(); ++it)
+    {
+      auto sortedIndex = std::distance(reordering.begin(), it);
+      // ROS_DEBUG_STREAM("origIndex: " << (*it).second << ", height: " << (*it).first << ", sortedIndex = " << sortedIndex);
+      unsorted.row((*it).second).copyTo(sorted.row(sortedIndex));
+      sorted_keys.push_back((*it).first);
+    }
+
+    return sorted;
```

Comment:
This function could be re-implemented with [`cv::sortIdx`](https://docs.opencv.org/2.4/modules/core/doc/operations_on_arrays.html?highlight=sortidx#void%20sortIdx%28InputArray%20src,%20OutputArray%20dst,%20int%20flags%29), then create some LookUpTable, instead of using `copyTo`, which is the expensive bit here